robotframework-ls/src/robotframework_ls/impl/flow_explorer_model_builder.py

```diff
@@ -1,9 +1,53 @@
+from contextlib import contextmanager
 from robotframework_ls.impl.protocols import ICompletionContext, IKeywordFound, INode
 from typing import List, Any, Dict, Tuple
 import os
 from robocorp_ls_core.basic import isinstance_name
 
 
+class _KeywordRecursionStack:
+    _stack: Dict
+
+    def __init__(self) -> None:
+        self._stack = {}
+
+    def has(self, keyword_name: str) -> bool:
+        return keyword_name in self._stack
+
+    def push(self, keyword_name: str) -> None:
+        self._stack.update({keyword_name: None})
+
+    def pop(self, keyword_name: str) -> None:
```

Comment:
Like `push`, `pop` should be protected i.e.: `_pop` (or it could just be inlined).